services/statestorage/merkle/merkle_forest_test.go

```diff
@@ -29,137 +27,68 @@ import (
-	require.Equal(t, updatedRoot, topRoot, "getTopRootHash did not match getRootHash with rootId %v", rootId)
-}
-
-func TestGetPastRootHash(t *testing.T) {
-	f := newForestDriver()
-
-	updateStringEntries(f, "first", "val")
-	topRootOf1, err1 := f.getTopRootHash()
-	updateStringEntries(f, "second", "val")
-	rootOfOneAfterSecondUpdate, err2 := f.getRootHash(1)
-
-	require.NoError(t, err1, "GetTopHash failed in first call")
-	require.NoError(t, err2, "GetTopHash failed in second call")
-	require.Equal(t, rootOfOneAfterSecondUpdate, topRootOf1, "getRootHash did not return expected hash")
-}
-
 func TestRootChangeAfterStateChange(t *testing.T) {
-	f := newForestDriver()
+	f, root := NewForest()
 
-	updateStringEntries(f, "first", "val")
-	topRootOf1, err1 := f.getTopRootHash()
-	updateStringEntries(f, "first", "val1")
-	topRootOf2, err2 := f.getTopRootHash()
+	root1 := updateStringEntries(f, root, "first", "val")
+	root2 := updateStringEntries(f, root1, "first", "val1")
 
-	require.NoError(t, err1, "GetTopHash failed in first call")
-	require.NoError(t, err2, "GetTopHash failed in second call")
-	require.NotEqual(t, topRootOf1, topRootOf2, "root hash did not change after state change")
+	require.NotEqual(t, root1, root2, "root hash did not change after state change")
 }
 
 func TestRevertingStateChangeRevertsMerkleRoot(t *testing.T) {
-	f := newForestDriver()
+	f, root := NewForest()
 
-	updateStringEntries(f, "first", "val")
-	topRootOf1, err1 := f.getTopRootHash()
-	updateStringEntries(f, "first", "val1")
-	updateStringEntries(f, "first", "val")
-	topRootOf3, err2 := f.getTopRootHash()
+	root1 := updateStringEntries(f, root, "first", "val")
+	root2 := updateStringEntries(f, root1, "first", "val1")
+	root3 := updateStringEntries(f, root2, "first", "val")
 
-	require.NoError(t, err1, "GetTopHash failed in first call")
-	require.NoError(t, err2, "GetTopHash failed in second call")
-	require.Equal(t, topRootOf1, topRootOf3, "root hash did not revert back after resetting state")
+	require.Equal(t, root1, root3, "root hash did not revert back after resetting state")
 }
 
 func TestValidProofForMissingKey(t *testing.T) {
-	f := newForestDriver()
+	f, root := NewForest()
 	key := "imNotHere"
 	contract := "foo"
-	proof := getProofRequireHeight(t, f, 0, contract, key, 1)
-	verifyProof(t, f, 0, proof, contract, key, "", true)
-	verifyProof(t, f, 0, proof, contract, key, "non-zero", false)
+	proof := getProofRequireHeight(t, f, root, contract, key, 1)
+	verifyProof(t, f, root, proof, contract, key, "", true)
+	verifyProof(t, f, root, proof, contract, key, "non-zero", false)
 
 }
 
 func TestAddSingleEntryToEmptyTree(t *testing.T) {
-	f := newForestDriver()
-	rootId := updateStringEntries(f, "bar", "baz")
-	require.EqualValues(t, 1, rootId, "unexpected root id")
+	f, root := NewForest()
+	topRoot := updateStringEntries(f, root, "bar", "baz")
 
-	getProofRequireHeight(t, f, rootId, "", "bar", 1)
+	getProofRequireHeight(t, f, topRoot, "", "bar", 1)
 }
 
```

Comment:
i think this test should be first one, and add require that topRoot != root

Reply:
moved to top